### PR TITLE
George/refactor

### PR DESCRIPTION
**Pull request by george43g**
- remove express app
- scedule daily backup for prod
- use pubsub to trigger function
- we could also run it directly from local env
- other small refactors

### CODE REVIEW

File: apps/backend-functions/src/main.ts [hunk 1]
```diff
@@ -246,10 +245,6 @@ export const onOrganizationDeleteEvent = onDocumentDelete(
...
-/**
- * Trigger: REST call to the /admin app
- *
- *  - Backups / Restore the database
- *  - Quorum Deploy & setup a movie smart-contract
- */
-export const admin = region('us-central1').runWith(heavyConfig).https.onRequest(adminApp);
+export { default as dailyFirestoreBackup } from './pubsub/daily-firestore-backup';
+
+export { default as firestorePubsub } from './pubsub/firestore';
```
Comment (phpgeekfr) [inline apps/backend-functions/src/main.ts:250]: can you add a comment to explain a bit ?

File: apps/backend-functions/src/pubsub/firestore.ts [hunk 2]
```diff
@@ -0,0 +1,34 @@
+import { db, functions } from '../internals/firebase'
+import { EventContext, RuntimeOptions } from 'firebase-functions';
+import { Message } from 'firebase-functions/lib/providers/pubsub';
+import { Bucket } from '@google-cloud/storage';
+import { clear, exportFirestoreToBucket, getBackupBucket, restoreFromBackupBucket } from '@blockframes/firebase-utils';
+
+const heavyConfig: RuntimeOptions = {
+  timeoutSeconds: 300,
+  memory: '1GB',
+};
```
Comment (phpgeekfr) [inline apps/backend-functions/src/pubsub/firestore.ts:10]: can you use the one from main.ts ?

File: apps/backend-ops/src/admin.ts [hunk 3]
```diff
@@ -3,88 +3,34 @@
...
 /**
  * Trigger a firestore database restore operation for the given project
  */
-export async function restore(appURL: string) {
-  if (process.env['ADMIN_PASSWORD'] === undefined) {
-    throw new Error('no ADMIN_PASSWORD in your env, we need this to trigger backups / restores');
+export async function restore() {
+  const pubsub = new PubSub({ projectId: firebase.projectId });
+  await pubsub.topic('firestore').publish(Buffer.from('import'));
+  await sleep(5000);
```
Comment (phpgeekfr) [inline apps/backend-ops/src/admin.ts:16]: why this sleep ?

File: apps/backend-ops/src/admin.ts [hunk 4]
```diff
@@ -3,88 +3,34 @@
...
+  for (let i = 0; i < 10; i++) {
+    const maintenance = await isInMaintenance(120 * 1000); // 2 minutes delay
+    if (!maintenance) {
+      console.log('Restore process ended!');
+      return;
+    } else {
+      console.log('Waiting for the restore process to end ...');
+      await sleep(1000 * 50);
+    }
   }
```
Comment (phpgeekfr) [inline apps/backend-ops/src/admin.ts:26]: This was a hack to be sure that restore was completed even if admin app responded with 302, do you still need this to check when restore is completed? Also, were is maintenance mode activated now ?

File: apps/backend-ops/src/admin.ts [hunk 5]
```diff
@@ -3,88 +3,34 @@
...
-      if (error) {
-        reject(error);
-      } else {
-        resolve(response);
-      }
-    });
-  });
+export async function backup() {
+  const pubsub = new PubSub({ projectId: firebase.projectId });
+  await pubsub.topic('firestore').publish(Buffer.from('export'));
```
Comment (phpgeekfr) [inline apps/backend-ops/src/admin.ts:35]: With this new way, how can we know that backup is ended?

Comment (phpgeekfr) [reply apps/backend-ops/src/admin.ts:35]: ADMIN_PASSWORD env var should be removed from sample.env

File: apps/backend-ops/src/admin.ts (hunk 4 above)
Comment (george43g) [reply apps/backend-ops/src/admin.ts:26]: Yes. Maintenance mode is activated by the backup/restore functions themselves.

File: apps/backend-ops/src/admin.ts (hunk 5 above)
Comment (george43g) [reply apps/backend-ops/src/admin.ts:35]: There are a couple of solutions here. but we never really could anyway... I'm not sure if you're supposed to wait on a https function to complete if it's running a long process.... In order to write a scalable and reliable completion detection, we would have to send a message back.
OR, we could write an actual interface in the CRM app (but we're basically re-creating the express admin app that we wanted to remove in the first place, it's a step backwards) that triggers a callable function and then watches the DB for a complete signal, or waits for the function to return. But yeah, we'll face the same timeout issues in that case.
The best way would be to either send a pubsub message back or raise a flag in the DB.

File: apps/backend-ops/src/admin.ts (hunk 3 above)
Comment (george43g) [reply apps/backend-ops/src/admin.ts:16]: wait for the function to cold start and enable maintenance mode in the first place before we check for it. not great.

File: apps/backend-functions/src/pubsub/firestore.ts (hunk 2 above)
Comment (george43g) [reply apps/backend-functions/src/pubsub/firestore.ts:10]: yeah I can. It should really be stored in a lib or abstracted away.

File: apps/backend-functions/src/main.ts (hunk 1 above)
Comment (george43g) [reply apps/backend-functions/src/main.ts:250]: The idea was to start doing things a bit more correctly. A function is named after its export key. Each function should exist in its own scope to speed up cold start. The whole main.ts needs to be rewritten... backend-functions would need to be. I'm not sure if you want to start doing things the right way or just stick to the old way until we re-write the whole thing.

Comment (george43g) [reply apps/backend-functions/src/main.ts:250]: I can add a comment to explain what these functions do

Comment (phpgeekfr) [reply apps/backend-functions/src/main.ts:250]: yeah just stick to the old way until we re-write the whole thing.

File: apps/backend-ops/src/admin.ts (hunk 4 above)
Comment (phpgeekfr) [reply apps/backend-ops/src/admin.ts:26]: for backup, we don't need to enable maintenance mode, only for restore

File: apps/backend-ops/src/admin.ts (hunk 5 above)
Comment (phpgeekfr) [reply apps/backend-ops/src/admin.ts:35]: ADMIN_PASSWORD is still there, can you remove all occurrences if not needed anymore ?

File: sample.env [hunk 6]
```diff
@@ -4,7 +4,6 @@
 
 ETHEREUM_MNEMONIC="";
 CASCADE8_ADMIN=""
```
Comment (phpgeekfr) [inline sample.env:6]: you have occurrences ADMIN_PASSWORD at bottom of file

Comment (phpgeekfr) [reply sample.env:6]: same in tools/scripts/deploy-function-config.ts and .circleci/deploy-and-test-e2e.sh

**Review by phpgeekfr — COMMENTED** [review]
when I run prepareForTesting (I deployed my functions before) :
```
DB backup in local storage bucket is being restored into Firestore...
Error: 7 PERMISSION_DENIED: User not authorized to perform this action.
    at Object.callErrorFromStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/call.ts:81:24)
    at Object.onReceiveStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/client.ts:334:36)
    at Object.onReceiveStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/client-interceptors.ts:434:34)
    at Object.onReceiveStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/client-interceptors.ts:397:48)
    at Http2CallStream.outputStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/call-stream.ts:230:22)
    at Http2CallStream.maybeOutputStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/call-stream.ts:280:14)
    at Http2CallStream.endCall (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/call-stream.ts:263:12)
    at Http2CallStream.handleTrailers (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/call-stream.ts:412:10)
    at ClientHttp2Stream.emit (events.js:315:20)
    at ClientHttp2Stream.EventEmitter.emit (domain.js:483:12) {
  code: 7,
  details: 'User not authorized to perform this action.',
  metadata: Metadata { internalRepr: Map {}, options: {} },
  note: 'Exception occurred in retry method that was not classified as transient'
}

```

**Comment by phpgeekfr** [comment]
> when I run prepareForTesting (I deployed my functions before) :
> 
> ```
> DB backup in local storage bucket is being restored into Firestore...
> Error: 7 PERMISSION_DENIED: User not authorized to perform this action.
>     at Object.callErrorFromStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/call.ts:81:24)
>     at Object.onReceiveStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/client.ts:334:36)
>     at Object.onReceiveStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/client-interceptors.ts:434:34)
>     at Object.onReceiveStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/client-interceptors.ts:397:48)
>     at Http2CallStream.outputStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/call-stream.ts:230:22)
>     at Http2CallStream.maybeOutputStatus (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/call-stream.ts:280:14)
>     at Http2CallStream.endCall (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/call-stream.ts:263:12)
>     at Http2CallStream.handleTrailers (/home/bruce/git/github/blockframes/node_modules/@grpc/grpc-js/src/call-stream.ts:412:10)
>     at ClientHttp2Stream.emit (events.js:315:20)
>     at ClientHttp2Stream.EventEmitter.emit (domain.js:483:12) {
>   code: 7,
>   details: 'User not authorized to perform this action.',
>   metadata: Metadata { internalRepr: Map {}, options: {} },
>   note: 'Exception occurred in retry method that was not classified as transient'
> }
> ```

Solved this by adding "pub sub administrator" role for my service account in https://console.cloud.google.com/iam-admin/

**Review by phpgeekfr — APPROVED** [review]

**Pull request merged**